shared/utils/usetaskOperations.ts

```diff
@@ -0,0 +1,157 @@
+import { useState, useEffect } from "react";
+import { Task } from "../../modules/HomePage/types/Task";
+
+const useTaskOperations = () => {
+  const [tasks, setTasks] = useState<Task[]>([]);
+  const [filter, setFilter] = useState<string>("all");
+  const [history, setHistory] = useState<Task[][]>([]);
+  const [redoArray, setRedoArray] = useState<Task[][]>([]);
+  const [editingTaskId, setEditingTaskId] = useState<string | null>(null);
+
+  useEffect(() => {
+    const savedTasks = window.localStorage.getItem("my-tasks");
+    if (savedTasks) {
+      const parsedTasks = JSON.parse(savedTasks);
+      setTasks(
+        parsedTasks.map((task: Task) => ({
+          ...task,
+          dueDate: new Date(task.dueDate),
+        }))
+      );
+    }
+  }, []);
+
+  useEffect(() => {
+    if (typeof window !== "undefined" && tasks.length > 0) {
+      window.localStorage.setItem("my-tasks", JSON.stringify(tasks));
+    }
+  }, [tasks]);
+
+  const updateTasks = (newTasks: Task[]) => {
+    setHistory((prevHistory) => [...prevHistory, tasks]);
+    setTasks(newTasks);
+    setRedoArray([]);
+  };
+
+  const addTask = (task: Task) => {
+    updateTasks([task, ...tasks]);
+  };
+
+  const deleteTask = (id: string) => {
+    updateTasks(tasks.filter((task) => task.id !== id));
+  };
+
+  const toggleCompletedTask = (id: string) => {
+    updateTasks(
+      tasks.map((task) =>
+        task.id === id ? { ...task, completed: !task.completed } : task
+      )
+    );
+  };
+
+  const deleteCompletedTasks = () => {
+    updateTasks(tasks.filter((task) => !task.completed));
+  };
+
+  const startEditingTask = (id: string) => {
+    setEditingTaskId(id);
+  };
+
+  const saveTask = (updatedTask: Task) => {
+    updateTasks(
+      tasks.map((task) => (task.id === updatedTask.id ? updatedTask : task))
+    );
+    setEditingTaskId(null);
+  };
+
+  const undo = () => {
+    setHistory((prevHistory) => {
+      if (prevHistory.length > 0) {
+        const lastState = prevHistory[prevHistory.length - 1];
+        setRedoArray((prevRedoState) => [...prevRedoState, tasks]);
+        setTasks(lastState);
+        return prevHistory.slice(0, prevHistory.length - 1);
+      }
+      return prevHistory;
+    });
+  };
+
+  const redo = () => {
+    setRedoArray((prevRedoState) => {
+      if (prevRedoState.length > 0) {
+        const nextState = prevRedoState[prevRedoState.length - 1];
+        setHistory((prevHistory) => [...prevHistory, tasks]);
+        setTasks(nextState);
+        return prevRedoState.slice(0, prevRedoState.length - 1);
+      }
+      return prevRedoState;
+    });
+  };
+
+  const applyFilter = (tasks: Task[], filter: string): Task[] => {
```

Comment:
This method is doing a lot of stuff together which makes it very difficult to read. Could you please simplify this method a bit. I suggest seperating 3 different filters in 3 different methods and then adjust the caller of this method according to the change.
for e.g: 

filterByStatus
filterByPriority
filterByDueDate